add extra properties to woocommerceanalytics events

haszari opened the pull request:
This PR adds some extra properties to `woocommerceanalytics` events. For example, these properties allow us to filter events on WooCommerce version.

#### Changes proposed in this Pull Request:
* Adds `url` property to all `woocommerceanalytics` events.
* Adds `woo_version` property to all `woocommerceanalytics` events.

pb0Spc-ou-p2

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
This adds features to the existing [`woocommerce-analytics`](https://github.com/Automattic/jetpack/tree/master/modules/woocommerce-analytics) module.

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* This requires Jetpack and WooCommerce – if necessary, install WooCommerce, set it up (use the onboarding wizard), and add some products. Use `Cash on delivery` payment method for easy test checkout - you can enable this in `WooCommerce > Settings > Payment`.
* The events are disabled for development sites. You can circumvent (hack) this by adding `return true;` at the start of [`Jetpack_WooCommerce_Analytics::shouldTrackStore()`](https://github.com/Automattic/jetpack/blob/master/modules/woocommerce-analytics/wp-woocommerce-analytics.php#L40).
- Open network tab in browser dev tools, search for `pixel.wp.com`.
- View a product on the front end of your site. You should see a request for `woocommerceanalytics_product_view` event. This should have new properties (parameters) `url` and `woo_version`, and the values should match the home url and Woo version (respectively).
- Add/remove products from cart, view cart / checkout, complete purchase - the relevant events should all include the new properties.

#### Proposed changelog entry for your changes:
* Added url and woo_version to WooCommerce Analytics events.

(Not sure if this needs a changelog entry.)

haszari committed:
add extra properties to woocommerceanalytics events:
- url - store/site home url
- woo_version - version of woo plugin installed on the site

zinigor reviewed (dismissed):
Works as expected, code looks fine!

jeherve commented on modules/woocommerce-analytics/classes/wp-woocommerce-analytics-universal.php:
```diff
@@ -103,6 +103,8 @@ public function loop_session_events() {
 								'pt': '" . esc_js( $product_details['type'] ) . "',
 								'ui': '" . esc_js( $this->get_user_id() ) . "',
+								'url': '" . esc_js( home_url() ) . "',
+								'woo_version': '" . esc_js( WC()->version ) . "',
```
Before we use it here and below, maybe we should check and ensure WC version is a valid object, like we did in #10538?

haszari replied:
I don't think that's necessary - in #10538 the check is on `WC()->session` before we call methods on `session`. In this case we are reading the value of `WC()->version` (e.g. string or worst case null/undefined) as a property. 

Note this analytics code is all [conditional on Woo being active](https://github.com/Automattic/jetpack/blob/master/modules/woocommerce-analytics/wp-woocommerce-analytics.php#L62), so `WC()` should be fine.

jeherve reviewed (changes requested):
Since we have data we want to send out for every event (such as blog ID, user ID, and now the site URL or the Woo version in this PR, I wonder if this may be a good time to extract all this into its own method to avoid duplicating code. what do you think?

haszari committed:
factor out common record_event from loop_session_events:
- prep work for having a single record event where we can add all
common properties
- also provides a PHP interface for logging events with custom params
+ tidy comments

haszari committed:
include common props js-triggered events via new util func

haszari committed:
use record_event for all remaining woo events

haszari committed:
fix issue with product_view event - handle no extra props

haszari commented:
> I wonder if this may be a good time to extract all this into its own method to avoid duplicating code. what do you think?

haszari commented:
This PR / branch has been rebased.

haszari commented:
Not sure if this needs GDPR review but added the tag just in case.

haszari commented:
FYI @psealock 👋

jeherve commented on modules/woocommerce-analytics/classes/wp-woocommerce-analytics-universal.php:
```diff
@@ -77,36 +77,67 @@ public function enqueue_tracking_script() {
+	 */
+	public function loop_session_events() {
+		// Check for previous events queued in session data.
 		if ( is_object( WC()->session ) ) {
```
Should we check for `WC()->version` here as well, since we'll use it later?

haszari replied:
Similar to [previous comment](https://github.com/Automattic/jetpack/pull/15028#discussion_r394652146), I don't think this is necessary.We're dereferencing `WC()->session`, e.g. `WC()->session->get( 'wca_session_data' );`. With `version` we're just adding the value as a param, so no biggie if it's missing (and it's better to have the `woo_version` prop consistently available on events, even if the value is `NULL`).

Happy to put the check in if it's needed - i.e. let me know if there's something I'm missing here.

jeherve replied:
Oh my bad, I missed your last comment!

jeherve commented on modules/woocommerce-analytics/classes/wp-woocommerce-analytics-universal.php:
```diff
@@ -223,22 +237,14 @@ public function checkout_process() {
+			);
 		}
-
-		wc_enqueue_js( implode( "\r\n", $universal_commands ) );
```
Do we need to include this in `record_event`?

haszari replied:
I don't think so - this was just rendering multiple `_wca.push()` calls (as JS strings) with whitespace in between.

In `record_event` we're enqueuing the `_wca.push()` call directly. 

The way to test this is to test an event that trigger on page load; for example add-to-cart from shop page, then view product page.

deltaWhiskey commented on modules/woocommerce-analytics/classes/wp-woocommerce-analytics-universal.php:
```diff
@@ -77,36 +77,67 @@ public function enqueue_tracking_script() {
+					'pp': '" . esc_js( $product_details['price'] ) . "',
+					'pt': '" . esc_js( $product_details['type'] ) . "'," .
+					$extra_properties . $this->get_common_properties() . '
+				} );'
```
We're doing a lot of string concatenation and repeated calls to `esc_json()`. That preserves the previous coding style, but I think it would be more readable and maintainable to pass arrays around, with the values unescaped.

When the string finally has to be generated (in `event_push()`, if we go with that approach, or else here,) that would be the moment to escape everything with an `array_walk()` call, and then encode with `json_encode()`. That would mean we need to review where single vs. double quotes are being used, wherever we output this.

If we were working with arrays before output, then `get_common_properties()` could be passing this a raw array, without doing part of the escaping there, and part in the various other places it gets called.

haszari replied:
I agree it would be nice to have a single path for all these events. The issue is that those two other events are generated on the front end, and pull data from the DOM, so they call the JS api directly - we can't enqueue a pre-rendered `_wca.push()` call.

In future, we could render a `event_push()` style wrapper func on the front end which supplies all the default props. I didn't want to go that far with the refactor :)

deltaWhiskey reviewed (commented):
It looks like it should work. It's worth making sure the endpoint is getting what's expected. If you'd like to work together on that, let me know.

I added some organizational / coding style suggestions.

psealock commented:
@haszari Can you make a list of events affected? I can go through them and compare to master to make sure no changes have been made to the properties already in production.

haszari commented:
> @haszari Can you make a list of events affected? I can go through them and compare to master to make sure no changes have been made to the properties already in production.

Yes please – I'd appreciate another test pass to make sure the events are still working correctly. I've tested and compared with master, but there are lots of props to eyeball. Thanks @psealock !

All 5 events are affected (I believe these are all the WCA events?):

```
woocommerceanalytics_product_view
woocommerceanalytics_add_to_cart
woocommerceanalytics_remove_from_cart
woocommerceanalytics_product_checkout
woocommerceanalytics_product_purchase
```

haszari committed:
refactor: split definition of common props from render as js object keys

haszari commented on modules/woocommerce-analytics/classes/wp-woocommerce-analytics-universal.php:
```diff
@@ -77,36 +77,67 @@ public function enqueue_tracking_script() {
 		$blogid = Jetpack::get_option( 'id' );
 
-		// check for previous add-to-cart cart events
+		return "'blog_id': '" . esc_js( $blogid ) . "',
```
Good point! What I've done is to split the definition of the common props and rendering (as js object props fragment) into separate methods (40e0171e7cfe408417133a5dc1d5613e1ae1febc).

It's still using the string-concatenation style, with embedded `esc_js` call, for consistency with the existing code (cc @deltaWhiskey). But now we can deal with common/server-side props with PHP arrays and then call `render_properties_as_js` to render them in a standard way.

haszari commented:
Thanks for the reviews & feedback – this PR is all the better for your feedback! Please give this a final review & test when you're ready. @jeherve @psealock @deltaWhiskey

haszari commented:
I see Travis is failing, I can't tell if this is due to issues caused by my code. 

There are two jobs failing:

- [PHP Nightly](https://travis-ci.org/github/Automattic/jetpack/jobs/665643904) seems to be something in photon, `test_photon_cdn_in_rest_response_with_edit_context` is failing
- [PHP 7.2](https://travis-ci.org/github/Automattic/jetpack/jobs/665643890) is timing out at 10 minutes

Please ping me / comment if there are steps I should take to resolve these!

jeherve reviewed (approved):
This looks good and tests well for me. 👍 I'll let someone more familiar with the events take a look at it as well, but on my end I think this should be good to merge.

As for Travis, it should be okay now.

haszari commented:
@psealock Can you give these events a final test/review before merge? Thanks :)

psealock commented on modules/woocommerce-analytics/classes/wp-woocommerce-analytics-universal.php:
```diff
@@ -76,37 +76,84 @@ public function enqueue_tracking_script() {
+		return array(
+			'blog_id'     => Jetpack::get_option( 'id' ),
+			'ui'          => $this->get_user_id(),
+			'url'         => home_url(),
```
I'm curious what the url is for. We already have browser document location in `_dl` property. Are you looking for a unique way to identify sites? If so, would `blog_id` suffice?

deltaWhiskey replied:
`home_url()` is to give a way to map to data collected by WC Tracker, which does not have `blog_id`. There have been experiments in the past with using document location, but that comes with some challenges.

psealock commented on modules/woocommerce-analytics/classes/wp-woocommerce-analytics-universal.php:
```diff
@@ -76,37 +76,84 @@ public function enqueue_tracking_script() {
+			'blog_id'     => Jetpack::get_option( 'id' ),
+			'ui'          => $this->get_user_id(),
+			'url'         => home_url(),
+			'woo_version' => WC()->version,
```
How about keeping this inline with other property names: `wv`?

haszari replied:
Prefer clarity over terseness here. See also @deltaWhiskey comment in support of clearer props names: https://github.com/Automattic/jetpack/pull/15028#discussion_r395233504

psealock reviewed (commented):
All the events check out, code looks good. Nice job @haszari 

I just have the two questions above, otherwise looks good.

psealock reviewed (approved):
Thanks for answering my questions. Code looks good and events are working well.

haszari commented:
@jeherve – I've merged this – I hope that's ok! Let me know if I need to do anything more to wrangle this change (e.g. changelog, other review tags).

jeherve commented:
@haszari You're all set, nothing else to do! Thank you!
